In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
x = lending_data_df.drop(columns=['loan_status'])

In [5]:
# Review the y variable Series
y.head(20)

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
x.head(20)

loan_size  interest_rate  borrower_income  debt_to_income  \
0     10700.0          7.672            52800        0.431818   
1      8400.0          6.692            43600        0.311927   
2      9000.0          6.963            46100        0.349241   
3     10700.0          7.664            52700        0.430740   
4     10800.0          7.698            53000        0.433962   
5     10100.0          7.438            50600        0.407115   
6     10300.0          7.490            51100        0.412916   
7      8800.0          6.857            45100        0.334812   
8      9300.0          7.096            47400        0.367089   
9      9700.0          7.248            48800        0.385246   
10     9300.0          7.085            47300        0.365751   
11     9200.0          7.015            46600        0.356223   
12     9200.0          7.043            46900        0.360341   
13     9400.0          7.101            47400        0.367089   
14     8400.0          6.703            43700        0.313501   
15     9400.0          7.136            47700        0.371069   
16     8900.0          6.897            45500        0.340659   
17     8300.0          6.648            43200        0.305556   
18     8000.0          6.539            42100        0.287411   
19    10200.0          7.460            50800        0.409449   

    num_of_accounts  derogatory_marks  total_debt  
0                 5                 1       22800  
1                 3                 0       13600  
2                 3                 0       16100  
3                 5                 1       22700  
4                 5                 1       23000  
5                 4                 1       20600  
6                 4                 1       21100  
7                 3                 0       15100  
8                 3                 0       17400  
9                 4                 0       18800  
10                3                 0       17300  
11                3                 0       16600  
12                3                 0       16900  
13                3                 0       17400  
14                3                 0       13700  
15                3                 0       17700  
16                3                 0       15500  
17                2                 0       13200  
18                2                 0       12100  
19                4                 1       20800

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(x, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)
print(len(testing_predictions))

19384


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
logistic_acc_score = balanced_accuracy_score(y_test, testing_predictions)
print(logistic_acc_score)

0.9442676901753825


In [12]:
# Generate a confusion matrix for the model
logistic_matrix = confusion_matrix(y_test, testing_predictions)
print(logistic_matrix)

[[18679    80]
 [   67   558]]


In [13]:
# Print the classification report for the model
logistic_classreport = classification_report(y_test, testing_predictions)
print(logistic_classreport)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** It accurately predicts healthy loans (0) with high precision and recall, as well as high-risk loans (1) with reasonably good precision and recall. The weighted average and overall accuracy of 0.99 further indicate the model's high performance.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [14]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
x_res, y_res = ros.fit_resample(X_train, y_train)

In [15]:
# Count the distinct values of the resampled labels data
y_res.value_counts()

0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [16]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
classifier.fit(x_res, y_res)

# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test})

Prediction  Actual
36831           0       0
75818           1       1
36563           0       0
13237           0       0
43292           0       0
...           ...     ...
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [17]:
# Print the balanced_accuracy score of the model 
predicted_logistic_acc_score = balanced_accuracy_score(y_test, predictions)
print(predicted_logistic_acc_score)

0.9959744975744975


In [18]:
# Generate a confusion matrix for the model
predicted_logistic_matrix = confusion_matrix(y_test, predictions)
print(predicted_logistic_matrix)

[[18668    91]
 [    2   623]]


In [19]:
# Print the classification report for the model
predicted_logistic_classreport = classification_report(y_test, predictions)
print(predicted_logistic_classreport)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model, with the data resampled, accurately predicts a healthy, low-risk loan with 100% precision. However, its precision in predicting high-risk loans is slightly lower at 87%. Overall, the model maintains a balanced accuracy of 100%.